src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+respectively). Patterns match recursively on their subpatterns.
+For example, the individual fields of any [collection-type][] pattern could be 
+[variable patterns][variable] or [constant patterns][constant]:
+
+```dart
+switch (obj) {       // List pattern [a,b] matches obj first if obj
+  case [a, b]: ...   // is a list wth two fields, then if its  
+}                    // fields match the constant subpatterns a and b  
+```
+
+Patterns can also _partially_ match, by using a [wildcard pattern][] as a place holder,
+or a [rest element][] in the case of list patterns.
+
+If you need to ensure `null` values aren't silently treated as match failures,
+use the [null-check pattern][] while matching.
+
+## Destructuring
+
+When an object and pattern match, the pattern can then access the object's data 
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+var [a, b, c] = numList;   // List pattern [a, b, c] destructures the list object
+print(a + b + c);          // numList into individual new parts. Prints 6.
+```
+
+You can use the [null-check pattern][] to check for `null` while destructuring. 
+
+All patterns attempt to match the objects they're compared against, but not all
+patterns destructure.
+Whether a pattern destructures depends on the [type of pattern][types]
+and the [context](#pattern-uses) in which it's being used.
+
+## Pattern uses
+
+Patterns integrate into several code constructs across the Dart language:
+
+- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
+- [for and for-in loops][for]
+- [if-cases] and [switch-cases]
+- [Collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns, 
+combining various contexts, pattern types, and surrounding code constructs.
+
+### Variable declaration
+
+Anywhere Dart allows regular, single-name local variable declaration, you can also
+use a _pattern variable declaration_. If the pattern matches the value on the right
+of the declaration, it destructures the value, binding it to new local variables:
+
+```dart
+var (a, [b, c]) = ("str", [1, 2]);   // Declares new variables a, b, and c.
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+Use the [null-assert pattern][] to eliminate `null` in variable declarations.
+
+### Variable assignment 
+
+A _variable assignment pattern_ on the left side of an assignment destructures the
+matched object and assigns the values to _existing_ variables,
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b);                   // Swap
+print('$a $b');                    // Prints "right left".
+```
+
+### Destrcuturing multiple returns
```

Comment:
This section is an outlier because the other sections are describing distinct places in the language grammar where you can use a pattern. This section is more like a use case or idiom for patterns.

Maybe instead of having this as a separate section, you could just use this code as an example in the pattern variable declaration section sort of like how you use swapping as an example idiom under pattern assignment?

Taking a step back, maybe that's a smoother way to integrate the "places you can use a pattern" and "common idioms for patterns" together? For each place you use can use a pattern, the idiom then serves as a concrete example. Something like:

* Pattern variable declarations -> Destructuring multiple returns
* Pattern assignment -> Swapping variables
* Pattern variables in for and for-in -> The `MapEntry(...)` example of using an object pattern for destructuring
* Switch statements and expressions -> Algebraic datatype-style code with exhaustiveness checking over a sealed class's subtypes

Reply:
This is MUCH better, will do!